lib/resonate.ts

```diff
@@ -160,100 +181,308 @@ export abstract class ResonateBase {
+   * Start the resonate service which continually checks for pending promises
+   * every `delay` ms.
    *
-   * @template T The return type of the function.
-   * @param id A unique id for the function invocation.
-   * @param name The function name.
-   * @param argsWithOpts The function arguments.
-   * @returns A promise that resolve to the function return value.
+   * @param delay Frequency in ms to check for pending promises.
    */
-  run<T>(nameOrTfc: string | TFC, id?: string, ...argsWithOpts: [...any, PartialOptions?]): ResonatePromise<T> {
-    let tfc!: TFC;
-    if (typeof nameOrTfc === "string") {
-      const { args, opts: givenOpts } = utils.split(argsWithOpts);
-      const { durable, eidFn, idempotencyKeyFn, retryPolicy, tags, timeout, version } = givenOpts;
-
-      if (!id) {
-        throw new Error("Id was not set for a top level function call");
-      }
-
-      tfc = {
-        funcName: nameOrTfc,
-        id,
-        args,
-        optsOverrides: {
-          durable,
-          eidFn,
-          idempotencyKeyFn,
-          retryPolicy,
-          tags,
-          timeout,
-          version,
-        },
-      };
-    } else {
-      tfc = nameOrTfc;
-    }
-
-    return this._run(tfc);
+  async start(delay: number = 5000) {
+    clearInterval(this.#interval);
+    this.#_start();
+    this.#interval = setInterval(this.#_start.bind(this), delay);
   }
 
-  private _run<T>(tfc: TFC): ResonatePromise<T> {
-    // Sets the defaults for the optional fields of TFC
-    tfc.optsOverrides = tfc.optsOverrides ?? {};
-    tfc.args = tfc.args ?? [];
-    tfc.optsOverrides.version = tfc.optsOverrides.version || 0;
+  /**
+   * Stop the resonate service.
+   */
+  async stop() {
+    // TODO: await all the invocation handles here
+    clearInterval(this.#interval);
+  }
 
-    if (!this.functions[tfc.funcName] || !this.functions[tfc.funcName][tfc.optsOverrides.version]) {
-      throw new Error(`Function ${tfc.funcName} version ${tfc.optsOverrides.version} not registered`);
+  async #_start() {
+    try {
+      for await (const promises of this.promisesStore.search("*", "pending", { "resonate:invocation": "true" })) {
+        for (const promiseRecord of promises) {
+          const param = this.defaultInvocationOptions.encoder.decode(promiseRecord.param.data);
+          if (
+            param &&
+            typeof param === "object" &&
+            "func" in param &&
+            typeof param.func === "string" &&
+            "version" in param &&
+            typeof param.version === "number" &&
+            "args" in param &&
+            Array.isArray(param.args) &&
+            "retryPolicy" in param &&
+            retryPolicies.isRetryPolicy(param.retryPolicy)
+          ) {
+            // Since the promise is already created on the server, we should use that idempotencyKey.
+            // If for whatever reason it is not, we should recalculate it using our defaults.
```

Comment:
It would be nice if we could "pass through" the durable promise and not have to call create again in `invokeLocal`. In the case where the idempotencyKeyForCreate has not been specified we will be unable to get past the create step.